pilot/pkg/networking/core/v1alpha3/cluster_builder.go

```diff
@@ -1143,15 +1142,24 @@ func (cb *ClusterBuilder) buildUpstreamClusterTLSContext(opts *buildClusterOpts,
 	return tlsContext, nil
 }
 
-func (cb *ClusterBuilder) setAutoSni(mc *MutableCluster, tls *networking.ClientTLSSettings) {
-	if mc != nil && features.EnableAutoSni && len(tls.Sni) == 0 {
-		if mc.httpProtocolOptions == nil {
-			mc.httpProtocolOptions = &http.HttpProtocolOptions{}
-		}
-		if mc.httpProtocolOptions.UpstreamHttpProtocolOptions == nil {
-			mc.httpProtocolOptions.UpstreamHttpProtocolOptions = &core.UpstreamHttpProtocolOptions{}
+// Set auto_sni if sni field is not explicitly set and if EnableAutoSni feature flag is enabled.
+// Set auto_san_validation if VerifyCertAtClient feature flag is enabled
+func (cb *ClusterBuilder) setAutoSniAndAutoSanValidation(mc *MutableCluster, tls *networking.ClientTLSSettings) {
+	if mc != nil && features.EnableAutoSni {
+		if len(tls.Sni) == 0 || (features.VerifyCertAtClient && len(tls.SubjectAltNames) == 0) {
+			if mc.httpProtocolOptions == nil {
+				mc.httpProtocolOptions = &http.HttpProtocolOptions{}
+			}
+			if mc.httpProtocolOptions.UpstreamHttpProtocolOptions == nil {
+				mc.httpProtocolOptions.UpstreamHttpProtocolOptions = &core.UpstreamHttpProtocolOptions{}
+			}
+			if len(tls.Sni) == 0 {
+				mc.httpProtocolOptions.UpstreamHttpProtocolOptions.AutoSni = true
+			}
+			if features.VerifyCertAtClient && len(tls.SubjectAltNames) == 0 {
```

Comment:
I think we don't need to check tls.SubjectAltNames, because if the TLS layer SAN checking fails, the connection can not setup , so it will not impact to L7 SAN checking. And if this TLS layer SAN checking passed, it also does not impact L7 checking. But if you would like to only keep one check, you can do that.  The pseudo-code may look like this:

```go
func setAutoSniAndAutoSANValidation(){
	if len(tls.Sni) != 0 || !features.EnableAutoSni
	    return
		
	if mc.httpProtocolOptions == nil {
		mc.httpProtocolOptions = &http.HttpProtocolOptions{}
	}
	if mc.httpProtocolOptions.UpstreamHttpProtocolOptions == nil {
		mc.httpProtocolOptions.UpstreamHttpProtocolOptions = &core.UpstreamHttpProtocolOptions{}
	}
        mc.httpProtocolOptions.UpstreamHttpProtocolOptions.AutoSni = true
	
	if features.VerifyCertAtClient //&& len(tls.SubjectAltNames) == 0
	    mc.httpProtocolOptions.UpstreamHttpProtocolOptions.AutoSanValidation = true
}
```
BTW, I also think AutoSanValidation is not related to VerifyCertAtClient.

Reply:
> 
auto_san_validation only enabled when auto_sni is enabled. or I miss anything here?
refer [here](https://www.envoyproxy.io/docs/envoy/v1.22.0/api-v3/config/core/v3/protocol.proto#envoy-v3-api-field-config-core-v3-upstreamhttpprotocoloptions-auto-san-validation)

Reply:
Auto sni or auto san validation can be be set only when they are not set in `UpstreamTlsContext`, otherwise it will override.

So here must check tls. SubjectAltNames and SNI

Reply:
@hobbytp envoy documentation says that auto_san_validation is "intended" to be set along with auto_sni as sni is needed for the validation. However since in istio we have a way to set SNI even otherwise through DR, I have just coded in such a way that if either auto_sni or explicit sni is present, we can use auto_san_validation, and that seem to work as well. Let me know what you think

Reply:
Consider one case:

In DestinationRule
    host: *.example.com
    SANs: [a.example.com, *.example.com]

Client (with sidecar) sends HTTP with host header= b.example.com
1) if server responses cert with SANs: [a.example.com, b.example.com, *.example.com] ==> L4 SAN check passes, and L7 SAN validation pass.
2) if server responses cert SANs: [a.example.com, *.example.com] ==> will L4 check pass? If yes, L7 SAN validation will fail for it can not find b.example.com in SANs.
This means: len(tls.SubjectAltNames) != 0, but it does not cover L7 logic checking.

So my question is why should we couple L4 SAN validation and L7 SAN validation even though they both are about tls handshake.

Reply:
The SAN check is not an `AND` operation - its an `OR`.

You should not be able to set explicitly SANs   and use auto san

Reply:
But I think we can, at least from the envoy proxy perspective, because one is for L4, one is for L7, and their configurations are separated, right?
But maybe I got your point, you want to disable the auto_san_validation (in L7) when L4 configured SAN list in DR from istio perspective? Just like you said "You should not be able to set explicitly SANs and use auto san", right?

Reply:
> 
@kfaseela  auto_san_validation is "intended" to be set with auto_sni field WHEN [override_auto_sni_header](https://www.envoyproxy.io/docs/envoy/v1.22.0/api-v3/config/core/v3/protocol.proto#envoy-v3-api-field-config-core-v3-upstreamhttpprotocoloptions-override-auto-sni-header) is unset. When override_auto_sni_header is set, auto_san_validation will compare the SAN value(s) from the upstream certificate with the override_auto_sni_header. The SNI value in override_auto_sni_header will override the original HTTP header host/authority value. 
The auto_san_validation is only used to compare the SAN from upstream certificate with the SNI value from client HTTP host/authority header or configured override_auto_sni_header.

Reply:
There is no L4  & L7, SAN check is during tls handshake. 

According to envoy document, auto_sni and auto_san_validation will override that specified in TransportSocketMatch, which is set from DR's TLS setting.

So here we definitely not to override user explicitly set ones.

Reply:
yes, I also found that now, you are right. 
So the left questions are which SNI, SAN we want to take high priority. DR's or auto.
seems both you and John's comments prefer DR's to be the high priority, right?

Reply:
Yep, we want to and must respect user explicitly setting, this is the convention of declaring API  Semantics.

Reply:
That sounds reasonable.